### PR TITLE
feat: Ability to set upper limit when using CPU auto-detection

### LINKED ISSUE #8279: Add `maxProcesses` option to `ParallelConfigFactory::detect()`
### Feature request

Hey.

We're using `ParallelConfigFactory::detect()` as it generally fits our needs well, aside from one issue: on some of our CI servers we have very high core counts (>100), and parallel detection detects them all and runs 100+ processes. Sure it's quick, but I'm sure dropping it down to 20 would be almost as fast, but without all the overhead. 

So it would be nice to have an option in the detect method to "clamp" the processes count to a specified maximum - e.g.:

```php
// Given there are 20 cores detected

ParallelConfigFactory::detect(); // uses 20 cores
ParallelConfigFactory::detect(maxProcesses: 30); // uses 20 cores
ParallelConfigFactory::detect(maxProcesses: 10); // clamps to 10 cores
```

I will PR the change :)

### Link to Discussion where feature request was formed

_No response_

### Contribution Checks

- [ ] I have verified if this feature request was already discussed
- [X] I am familiar with ["Feature or bug?"](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/feature-or-bug.rst)

### PR DESCRIPTION
**Pull request by oprypkhantc**
Closes #8279 

Feel free to make edits on the branch if you want :)

### CODE REVIEW

**Comment by Wirone** [comment]
@theofidry Sorry for pinging, but I would like to discuss this with you. Does it make sense to implement it on Fixer's side? Maybe your `CpuCoreCounter` should allow providing upper CPU limit?

**Comment by oprypkhantc** [comment]
Cool :) Changed it to use the new method, and added 1 core as reserved for the main process. I'm guessing there's no point in exposing the `$reservedCores` parameter to users?

**Comment by Wirone** [comment]
> I'm guessing there's no point in exposing the $reservedCores parameter to users?

You're right. From Fixer's perspective we should reserve 1 process for the main command and only detect how many child processes we can spawn (with an upper limit provided by a user).

**Review by Wirone — APPROVED** [review]

**Comment by Wirone** [comment]
@oprypkhantc it's OK, but you need to fix the merge conflicts first (merge or rebase on latest `master`) 🙂.

File: src/Runner/Parallel/ParallelConfigFactory.php [hunk 1]
```diff
@@ -47,8 +49,11 @@ public static function detect(
             ]);
         }
 
+        // Reserve 1 core for the main orchestrating process
+        $available = self::$cpuDetector->getAvailableForParallelisation(1, $maxProcesses);
```
Comment (theofidry) [inline src/Runner/Parallel/ParallelConfigFactory.php:53]: is this really necessary? Once you stat the parallel processes, the main process is likely gonna sleep

**Comment by Wirone** [comment]
Thank you @oprypkhantc 🍻!

**Pull request merged**